protocol/opcm-test-integration.md

```diff
@@ -0,0 +1,130 @@
+# Direction for OPCM test integration
+
+## Purpose
+
+In order to integrate the OPCM into the forge test setup (ie. `Deploy.s.sol`), we must first
+simplify the deploy scripts while keeping the desired properties and removing what we don't need.
+
+## Goals
+
+As far as we know, these scripts are not currently being used by chain operators for deployment, and
+we intend to do that with OPCM in the future anyways. Therefore, what we need to maintain about the
+existing flow is:
+- Artifact generation:
+  - Contracts should have the same names in the resulting deploy artifacts file at
+    `/deployments/{chainId}-deploy.json`
+- EVM state:
+  - The test values configured by each test suite must be preserved (otherwise the diff to update
+    each test will explode).
+
+Nice to haves:
+- logs emitted during deployment:
+  - This is not essential, but comparing before and after logs is a quick and dirty way to ensure
+    that the config is roughly the same.
+- Deploying contracts to the same address as before.
+  - This might be achievable based on the use of
+    [`create2`](https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/scripts/deploy/Deploy.s.sol#L1703)
+    in the existing scripts. If so it would enable us to use statedumps to diff the before and after
+    deployments.
+
+## Proposed Solution
+
+A plausible PR stack for achieving this might be:
+
+1. Remove the `viaSafe` functionality: `Deploy.s.sol` currently starts our by deploying a new
+  `SystemOwnerSafe`, and then uses `callViaSafe()` and `upgradeAndCallViaSafe()` for all calls to
+    the `ProxyAdmin` While maintaining the `viaSafe()` helper methods (they are also needed in
+  `DeployOwnership` for now), we should remove all usages in `Deploy.s.sol`, perform all actions
+    through the deployer key, then finally transfer the `ProxyAdmin` to the `finalSystemOwner`
+    (which the current scripts don't even do at the moment).
+2. Deploy a separate `ProxyAdmin` contract for the Superchain. A key diff between the OPCM approach
+    and `Deploy.s.sol` is that the latter shares the same `ProxyAdmin` for the Superchain and the OP
+    Chain that it deploys. That needs to be fixed to prepare for the OPCM.
+3. Deploy the Superchain contracts using the OPCM's `DeploySuperchain.s.sol`. With the previous PRs,
+    this will hopefully be straight-forward.
+4. Deploy the OP Chain Contracts using the `OPContractsManager` contract. This will be much simpler
```

Comment:
Is there a missing `DeployImplementations` step in between here?

Reply:
Yes, good call out. I'm admittedly skipping a bit of detail by step 4 here. Will add that in.

Reply:
8d9901144613